### PR TITLE
SiteManager Dashboard: Add API endpoint for Data Corrections Tool

### PR DESCRIPTION
**Pull request by abhinavjonnada82**
Title^^

This PR addresses following issue: https://github.com/episphere/connect/issues/810

Related PR: https://github.com/episphere/dashboard/pull/599

### CODE REVIEW

File: utils/firestore.js [hunk 1]
```diff
@@ -3033,6 +3033,32 @@ const writeCancerOccurrences = async (cancerOccurrenceArray) => {
...
+        const docId = snapshot.docs[0].id;
+        delete  participantData['token']
+
+        if (participantData['state.148197146'] === 'NULL') {
+            delete participantData['state.148197146']
+            await db.collection("participants").doc(docId).update({
+                'state.148197146': admin.firestore.FieldValue.delete()
+            });
+        }
+        else if (Object.keys(participantData).length > 0) { // performs an update only if other key/value exists
```
Comment (Gbarra9) [inline utils/firestore.js:3049]:
```suggestion
        } else if (Object.keys(participantData).length > 0) { // performs an update only if other key/value exists
```

File: utils/firestore.js [hunk 2]
```diff
@@ -3033,6 +3033,32 @@ const writeCancerOccurrences = async (cancerOccurrenceArray) => {
 
 
 
+const updateParticipantCorrection = async (participantData) => {
+    try{
```
Comment (Gbarra9) [inline utils/firestore.js:3037]:
```suggestion
    try {
```

File: utils/firestore.js [hunk 3]
```diff
@@ -3033,6 +3033,32 @@ const writeCancerOccurrences = async (cancerOccurrenceArray) => {
...
+            });
+        }
+        else if (Object.keys(participantData).length > 0) { // performs an update only if other key/value exists
+            await db.collection("participants").doc(docId).update(
+                {...participantData}
+            )
+        }
+        return true;
+    }
+    catch(error){
```
Comment (Gbarra9) [inline utils/firestore.js:3056]:
```suggestion
    } catch (error) {
```

**Review by Gbarra9 — APPROVED** [review]
Looks good.

File: utils/sites.js [hunk 4]
```diff
@@ -478,8 +478,36 @@ const updateUserAuthentication = async (req, res, authObj) => {
...
+    if (req.method === 'OPTIONS') return res.status(200).json({ code: 200 });
+
+    if (req.method !== 'POST') {
+        return res.status(405).json(getResponseJSON('Only POST requests are accepted!', 405));
+    }
+
+    try {
+        const { updateParticipantCorrection } = require('./firestore');
+        if (req.body.data) {
+            const status = await updateParticipantCorrection(req.body.data[0]);
```
Comment (anthonypetersen) [inline utils/sites.js:494]: can this throw an error? i.e. data exists but doesn't have length?

File: utils/firestore.js [hunk 5]
```diff
@@ -3033,6 +3033,30 @@ const writeCancerOccurrences = async (cancerOccurrenceArray) => {
...
+const updateParticipantCorrection = async (participantData) => {
+    try {
+        const snapshot = await db.collection('participants').where('token', '==', participantData['token']).get();
+        if (snapshot.empty) return false
+        const docId = snapshot.docs[0].id;
+        delete  participantData['token']
+
+        if (participantData['state.148197146'] === 'NULL') {
+            delete participantData['state.148197146']
+            await db.collection("participants").doc(docId).update({
```
Comment (anthonypetersen) [inline utils/firestore.js:3045]: single quotes around participants for consistency

File: utils/firestore.js [hunk 6]
```diff
@@ -3033,6 +3033,30 @@ const writeCancerOccurrences = async (cancerOccurrenceArray) => {
...
+                'state.148197146': admin.firestore.FieldValue.delete()
+            });
+        } else if (Object.keys(participantData).length > 0) { // performs an update only if other key/value exists
+            await db.collection("participants").doc(docId).update(
```
Comment (anthonypetersen) [inline utils/firestore.js:3049]: single quotes for consistency

File: utils/sites.js [hunk 7]
```diff
@@ -478,8 +478,36 @@ const updateUserAuthentication = async (req, res, authObj) => {
...
+            return status === true
```
Comment (anthonypetersen) [inline utils/sites.js:495]: do we need to check that status === true?

**Review by anthonypetersen — COMMENTED** [review]
comments left

**Review by anthonypetersen — APPROVED** [review]

**Pull request merged**